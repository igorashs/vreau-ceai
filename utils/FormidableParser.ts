import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
