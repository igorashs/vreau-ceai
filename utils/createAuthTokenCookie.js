import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
