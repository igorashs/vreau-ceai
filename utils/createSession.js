import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
