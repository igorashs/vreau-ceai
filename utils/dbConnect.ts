import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
