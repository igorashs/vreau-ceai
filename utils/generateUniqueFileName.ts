import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
