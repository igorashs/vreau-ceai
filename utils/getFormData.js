export const getFormData = (data) => {
  const formData = new FormData();

  for (const name in data) {
    if (name === 'src') continue;

    formData.append(name, data[name]);
  }

  if (data.src[0]) {
    formData.append('src', data.src[0]);
  }

  return formData;
};
