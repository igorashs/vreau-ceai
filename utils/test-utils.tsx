import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
