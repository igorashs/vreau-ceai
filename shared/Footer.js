import styled from 'styled-components';
import { Container } from './Container';
import breakpoints from 'GlobalStyle/breakpoints';
import { footerBlockLinksList } from '@/utils/links';
import { StyledLink } from '@/shared/StyledLink';

const Wrapper = styled.div`
  margin-top: calc(var(--baseline) * 2);
  padding: var(--baseline) 0;
  color: var(--text-light);
  background-color: var(--layout-dark);

  @media (min-width: ${breakpoints.lg}) {
    margin-top: calc(var(--baseline) * 4);
    padding: calc(var(--baseline) * 2) 0;
  }
`;

const BlockLinkList = styled.ul`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: calc(var(--baseline) / 2);

  @media (min-width: ${breakpoints.lg}) {
    display: grid;
    align-items: start;
    grid-template-columns: repeat(3, 224px);
    justify-content: space-between;
  }
`;

const LinkList = styled.ul`
  @media (min-width: ${breakpoints.lg}) {
    display: flex;
    flex-direction: column;
    gap: 7px;
  }
`;

const LinkListHeaderItem = styled.li`
  @media (min-width: ${breakpoints.lg}) {
    border-bottom: 1px solid #fff;
    padding-bottom: 7px;
  }
`;

const LinkListItem = styled.li`
  display: none;

  @media (min-width: ${breakpoints.lg}) {
    display: initial;
  }
`;

const CopyRight = styled.div`
  margin-top: var(--baseline);
  display: flex;
  gap: var(--baseline);
  justify-content: center;
  text-align: center;
  font-size: 0.75rem;

  @media (min-width: ${breakpoints.lg}) {
    margin-top: calc(var(--baseline) * 2);
  }
`;

const AccentWord = styled.span`
  color: var(--accent-text-light);
`;

const Divider = styled.span`
  width: 1px;
  background-color: var(--layout-light);
`;

export function Footer() {
  return (
    <Wrapper>
      <Container>
        <footer>
          <BlockLinkList>
            {footerBlockLinksList.map(({ href, text, label, links }) => (
              <li key={href}>
                <LinkList>
                  <LinkListHeaderItem>
                    <StyledLink href={href} text={text} label={label} />
                  </LinkListHeaderItem>
                  {links.map((link) => (
                    <LinkListItem key={link.href}>
                      <StyledLink {...link} />
                    </LinkListItem>
                  ))}
                </LinkList>
              </li>
            ))}
          </BlockLinkList>
          <CopyRight>
            <StyledLink
              href="https://github.com/igorashs/vreau-ceai"
              text="@Igorash"
              label="Visit Project Repo"
              target="_blank"
              rel="noreferrer"
            />
            <Divider />
            <p>
              made by <AccentWord>tea</AccentWord> nature
            </p>
          </CopyRight>
        </footer>
      </Container>
    </Wrapper>
  );
}
