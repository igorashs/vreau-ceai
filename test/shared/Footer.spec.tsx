import { render } from '@/utils/test-utils';
import Footer from '@/shared/Footer';

describe('Footer', () => {
  it('renders correctly', () => {
    const { container } = render(<Footer />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
