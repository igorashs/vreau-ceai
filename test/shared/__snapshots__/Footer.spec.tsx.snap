// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer renders correctly 1`] = `
.c5 {
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 7px;
  color: var(--text-light);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5:hover {
  -webkit-filter: brightness(0.7);
  filter: brightness(0.7);
}

.c7 {
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 7px;
  color: var(--accent-text-light);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7:hover {
  -webkit-filter: brightness(0.7);
  filter: brightness(0.7);
}

.c0 {
  margin-top: calc(var(--baseline) * 2);
  padding: var(--baseline) 0;
  color: var(--text-light);
  background-color: var(--layout-dark);
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: calc(var(--baseline) / 2);
}

.c6 {
  display: none;
}

.c8 {
  margin-top: var(--baseline);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: var(--baseline);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  font-size: 0.75rem;
}

.c10 {
  color: var(--accent-text-light);
}

.c9 {
  width: 1px;
  background-color: var(--layout-light);
}

@media (min-width:576px) {
  .c1 {
    width: 540px;
    max-width: 540px;
    margin: 0 auto;
  }
}

@media (min-width:768px) {
  .c1 {
    width: 720px;
    max-width: 720px;
  }
}

@media (min-width:992px) {
  .c1 {
    width: 960px;
    max-width: 960px;
  }
}

@media (min-width:1200px) {
  .c1 {
    width: 1140px;
    max-width: 1140px;
  }
}

@media (max-width:576px) {
  .c1 {
    width: 100%;
    padding: 0 15px;
  }
}

@media (min-width:992px) {
  .c0 {
    margin-top: calc(var(--baseline) * 4);
    padding: calc(var(--baseline) * 2) 0;
  }
}

@media (min-width:992px) {
  .c2 {
    display: grid;
    -webkit-align-items: start;
    -webkit-box-align: start;
    -ms-flex-align: start;
    align-items: start;
    grid-template-columns: repeat(3,224px);
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }
}

@media (min-width:992px) {
  .c3 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    gap: 7px;
  }
}

@media (min-width:992px) {
  .c4 {
    border-bottom: 1px solid #fff;
    padding-bottom: 7px;
  }
}

@media (min-width:992px) {
  .c6 {
    display: initial;
  }
}

@media (min-width:992px) {
  .c8 {
    margin-top: calc(var(--baseline) * 2);
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <footer>
      <ul
        class="c2"
      >
        <li>
          <ul
            class="c3"
          >
            <li
              class="c4"
            >
              <a
                aria-label="Deschideți pagina cu informații despre noi"
                class="c5"
                href="/about"
                rel=""
                target=""
              >
                Despre
              </a>
            </li>
            <li
              class="c6"
            >
              <a
                aria-label="Deschideți pagina cu informații despre noi"
                class="c7"
                href="/about"
                rel=""
                target=""
              >
                vreau ceai
              </a>
            </li>
          </ul>
        </li>
        <li>
          <ul
            class="c3"
          >
            <li
              class="c4"
            >
              <a
                aria-label="Deschideți pagina cu întrebări frecvente"
                class="c5"
                href="/faq"
                rel=""
                target=""
              >
                FAQ
              </a>
            </li>
            <li
              class="c6"
            >
              <a
                aria-label="Deschideți pagina cu întrebarea cum comand"
                class="c7"
                href="/faq#cum-comand"
                rel=""
                target=""
              >
                cum comand
              </a>
            </li>
            <li
              class="c6"
            >
              <a
                aria-label="Deschideți pagina cu întrebarea cum achit"
                class="c7"
                href="/faq#cum-achit"
                rel=""
                target=""
              >
                cum achit
              </a>
            </li>
            <li
              class="c6"
            >
              <a
                aria-label="Deschideți pagina cu întrebarea despre cum are loc livrarea"
                class="c7"
                href="/faq#livrare"
                rel=""
                target=""
              >
                cum are loc livrarea
              </a>
            </li>
          </ul>
        </li>
        <li>
          <ul
            class="c3"
          >
            <li
              class="c4"
            >
              <a
                aria-label="Deschideți pagina cu contacte"
                class="c5"
                href="/contacts"
                rel=""
                target=""
              >
                Contacte
              </a>
            </li>
            <li
              class="c6"
            >
              <a
                aria-label="apelați nr de tel"
                class="c7"
                href="tel:062222222"
                rel=""
                target=""
              >
                <svg />
                062222222
              </a>
            </li>
            <li
              class="c6"
            >
              <a
                aria-label="trimiteți email"
                class="c7"
                href="mailto:vreauceai@gmail.com"
                rel=""
                target=""
              >
                <svg />
                vreauceai@gmail.com
              </a>
            </li>
          </ul>
        </li>
      </ul>
      <div
        class="c8"
      >
        <a
          aria-label="Visit Project Repo"
          class="c5"
          href="https://github.com/igorashs/vreau-ceai"
          rel="noreferrer"
          target="_blank"
        >
          @Igorash
        </a>
        <span
          class="c9"
        />
        <p>
          made by 
          <span
            class="c10"
          >
            tea
          </span>
           nature
        </p>
      </div>
    </footer>
  </div>
</div>
`;
